File: day1/Solve.java
import java.util.ArrayList;
import java.util.Scanner;
import java.io.File;

import java.util.regex.Matcher;
import java.util.regex.Pattern;



public class Solve {
    public static void main(String[] args) {
        ArrayList<Line> lines = new ArrayList<>();
        try {
            File inputFile = new File("./day1/input.txt");
            Scanner inputFileReader = new Scanner(inputFile);
            while (inputFileReader.hasNextLine()) {
                lines.add(new Line(inputFileReader.nextLine()));
            }
            inputFileReader.close();
        } catch (Exception e) {
            System.out.println("An error occurred.");
            e.printStackTrace();
        }

        int part1 = 0;
        int part2 = 0;

        for (Line line : lines) {
            part1 += line.getValue();
            part2 += line.getPart2().getValue();
        }

        System.out.printf("Part 1: %d\n", part1);
        System.out.printf("Part 2: %d\n", part2);
    }
}

class Line {
    private String line;

    public Line(String line) {
        this.line = line;
    }

    public Line clone() {
        return new Line(line);
    }

    public Line getPart2() {
        String part2Line = line;
        Pattern pattern = Pattern.compile("one|two|three|four|five|six|seven|eight|nine");
        while (true) {
            Matcher matcher = pattern.matcher(part2Line);
            if (!matcher.find()) {
                break;
            }

            int index = matcher.start();
            String number = matcher.group();

            String before = part2Line.substring(0, index);
            String after = part2Line.substring(index + number.length());

            switch (number) {
                case "one":
                    part2Line = before + "1ne" + after;
                    break;
                case "two":
                    part2Line = before + "2wo" + after;
                    break;
                case "three":
                    part2Line = before + "3hree" + after;
                    break;
                case "four":
                    part2Line = before + "4our" + after;
                    break;
                case "five":
                    part2Line = before + "5ive" + after;
                    break;
                case "six":
                    part2Line = before + "6ix" + after;
                    break;
                case "seven":
                    part2Line = before + "7even" + after;
                    break;
                case "eight":
                    part2Line = before + "8ight" + after;
                    break;
                case "nine":
                    part2Line = before + "9ine" + after;
                    break;
            }
        }
        
        return new Line(part2Line);
    }

    public int getValue() {
        char[] chars = line.toCharArray();
        int first = -1;
        int last = -1;

        for (int i = 0; i < chars.length; i++) {
            if (chars[i] >= '0' && chars[i] <= '9') {
                if (first == -1) {
                    first = chars[i] - '0';
                }
                last = chars[i] - '0';
            }
        }

        return first * 10 + last;
    }
}

File: day1/Solve.cs
public class Solve
{
    static void Main(string[] args)
    {
        string[] strLines = System.IO.File.ReadAllLines("./day1/input.txt");
        Line[] lines = new Line[strLines.Length];
        for (int i = 0; i < strLines.Length; i++) lines[i] = new Line(strLines[i]);

        int part1 = 0;
        int part2 = 0;

        foreach (Line line in lines)
        {
            part1 += line.Value();
            part2 += line.Part2().Value();
        }

        System.Console.Write("Part 1:");
        System.Console.WriteLine(part1);
        System.Console.Write("Part 2:");
        System.Console.WriteLine(part2);
    }


}

class Line
{
    private readonly string line;

    public Line(string line)
    {
        this.line = line;
    }

    public int Value() {
        int first = -1;
        int last = -1;
        for (int i = 0; i < line.Length; i++)
        {
            if ('0' <= line[i] && line[i] <= '9')
            {
                if (first == -1)
                {
                    first = line[i] - '0';
                }
                last = line[i] - '0';
            }
        }
        return first * 10 + last;
    }

    public Line Part2()
    {
        string part2Line = line;
        for (int i = 0; i < part2Line.Length; i++)
        {
            bool is_1 = part2Line.Substring(i).StartsWith("one");
            bool is_2 = part2Line.Substring(i).StartsWith("two");
            bool is_3 = part2Line.Substring(i).StartsWith("three");
            bool is_4 = part2Line.Substring(i).StartsWith("four");
            bool is_5 = part2Line.Substring(i).StartsWith("five");
            bool is_6 = part2Line.Substring(i).StartsWith("six");
            bool is_7 = part2Line.Substring(i).StartsWith("seven");
            bool is_8 = part2Line.Substring(i).StartsWith("eight");
            bool is_9 = part2Line.Substring(i).StartsWith("nine");

            int value = 0
                + (is_1 ? 1 : 0)
                + (is_2 ? 2 : 0)
                + (is_3 ? 3 : 0)
                + (is_4 ? 4 : 0)
                + (is_5 ? 5 : 0)
                + (is_6 ? 6 : 0)
                + (is_7 ? 7 : 0)
                + (is_8 ? 8 : 0)
                + (is_9 ? 9 : 0);

            if (value != 0)
            {
                part2Line = part2Line.Substring(0, i) + value.ToString() + part2Line.Substring(i + 1);
            }
        }
        return new Line(part2Line);
    }
}